Fixes #1: Create mpy.sampler.BaseSampler

akhilpandey95 opened the pull request:
- created a BaseSampler class that takes an expression, distribution, and random_state as input parameters
- ensured to collect the distribution from the _dist attribution of every variable expression
- implemented a sample(self, size=None) method that returns a numpy array of (size, samples)
- moved the _sample() function inside the VarExpression(s) to the BaseSampler class

Signed-off-by: Akhil Akella <aakella@gpu3.lcrc.anl.gov>

akhilpandey95 committed:
Issue #1: Create mpy.sampler.BaseSampler

- created a BaseSampler class that takes an expression, distribution, and random_state as input parameters
- ensured to collect the distribution from the _dist attribution of every variable expression
- implemented a sample(self, size=None) method that returns a numpy array of (size, samples)
- moved the _sample() function inside the VarExpression(s) to the BaseSampler class

Signed-off-by: Akhil Akella <aakella@gpu3.lcrc.anl.gov>

akhilpandey95 committed:
Issue #1: Craete mpy.sampler.BaseSampler

- ensured List expressions are supported for BaseSampler
- created tests/test_sampler.py and added unittest file to test the BaseSampler
- removed commented out sampler tests code from tests/test_var.py

Signed-off-by: Akhil Akella <aakella@gpu3.lcrc.anl.gov>

akhilpandey95 committed:
Issue #1: Create mpy.sampler.BaseSampler

- added a mpy.BaseSampler._map_dist() function that maps a given
distribution map

Signed-off-by: Akhil Akella <aakella@gpu3.lcrc.anl.gov>

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
         self.rng = rng
+        self._cache = {}
+
+    def sample(self, size=None, to_nd_array=True):
```
replace `to_nd_array` with `flat`:
* if `flat=True` then return a flattened version of the samples (as a numpy array with 2 dimensions: `(size, num_variables)`) where all variables of the Expression are represented as columns.
* if `flat=False` then a list of dict (keys are variables id, values are variable samples) is returned.
The version `flat=True` should be faster in term of computation: when it is possible sample `size` samples by calling the `rvs` function once!.
Probably the variables should have an additional parameter/attribute named `slack` for example `mpy.List(["relu", "sigmoid"], slack_value="NA")` which is a "special" value when the variable is inactive by default it could be `"NA"` for `List`, `nan` for `Int` and `Float`

akhilpandey95 replied:
That makes sense. In essence I had the same concern about computational efficiency, as to what would be an efficient way to return the samples if it categorical variables are passed because we can't return an nd-array. I'll make this change.

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
+
+        # init a random state is rng
+        if self.rng is None:
+            self.rng = np.random.RandomState()
```
this if statement should be in the `__init__`

akhilpandey95 replied:
yeah

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
 
+        # check if the size is None
+        if size is None:
+            size = 1
```
we follow the same specification as `scipy` therefore `size=None` is not equivalent to `size=1`, when `size=int > 0` it returns a list or array of length `size`. when `size=None` it returns a `dict` if (`flat=False`) and ` 1-d array if `flat=True`.

akhilpandey95 replied:
okay

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
+            size = 1
 
+        # gather the list of choices
+        choices = self.exp.choices()
```
to avoid recomputing `choices` this could be performed once and stored in a `self._chocies` attribute

akhilpandey95 replied:
makes sense.

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
+            if to_nd_array:
+                # convert the samples to an nd_array
+                samples = np.array(samples).reshape(len(samples), -1)
+        except ValueError:
```
this should not happen (see previous comments)

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
+            # check if the expression is Float
+            if isinstance(var_exp, Float):
+                # add samples
+                s.append(self.distributions[var_id].rvs(loc=var_exp._low, \
```
this mechanic will break if a user decide to pass an other distribution, for example normal distribution

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
+            # check if the expression is List
+            if isinstance(var_exp, List):
+                # get idx of the samples to choice
+                var_exps_idx = self.rng.choice(var_exp._length(), size=1)
```
needs to be a distribution from https://docs.scipy.org/doc/scipy/reference/stats.html

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
+        # check if the distributions are passed
+        if distributions is None:
+            dist_map = {var_id: var_exp._dist if isinstance(var_exp, Int) or \
+                                  isinstance(var_exp, Float) else None \
```
`List` should also have ` default `_dist`

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
+            dist_map = {var_id: var_exp._dist if isinstance(var_exp, Int) or \
+                                  isinstance(var_exp, Float) else None \
+                                  for var_id, var_exp in self.exp.choices().items()}
+        else:
```
in this code, either the user as 0 default distribution or it has to define all of them, the user is allowed to pass only a few distributions. Also, this code will not accept multi-variate distributions where a key can be a "tuple" of variable names.

Deathn0t commented on tests/test_sampler.py:
```diff
@@ -0,0 +1,38 @@
+        samples = s.sample(size=size)
+
+        assert samples.shape[0] == size
+        assert samples.shape[1] == len(program.exp().variables().keys())
```
I don't remember, what is `.exp()` ?

Deathn0t commented on tests/test_sampler.py:
```diff
@@ -0,0 +1,38 @@
+        assert samples.shape[0] == 1
+        assert samples.shape[1] == len(a.exp().variables().keys())
+
+
```
tests are not sophisticated enough, many important test cases are missing
* nested variables
* more complex expressions
* mix of variables
* order of the columns of the array
* etc...

Deathn0t reviewed (commented):
Hello @akhilpandey95 I left some comments, it needs some additional work, please take a look at them. If you have questions do not hesitate.

akhilpandey95 committed:
Issue #1: Creates mpy.sampler.BaseSampler

- added _dist parameter for List(VarExpression) class
- updated sample(), _sample(), and _map_dist() methods
- minor modification to the test cases in tests/test_sampler.py

akhilpandey95 committed:
Issue #1: Creates mpy.sampler.BaseSampler

- adds tests for checking complex expressions, meta expressions, flat flag of the sample method, distribution map
- fixed the sample() and _map_dist() methods in BaseSampler to map distributions
- fixed the expected return types for the sample() method with various use cases of size, and flat parameter arguments

Signed-off-by: Akhil Akella <aakella@gpu3.lcrc.anl.gov>

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,12 +39,122 @@ def list_type(list_exp):
+
+                # check if the expression is Int
+                if isinstance(var_exp, Int):
+                    self._cache[var_id] = self._dist_map[var_id](low=var_exp._low, \
```
don't put ` \ ` when it is not necessary

akhilpandey95 committed:
Issue #1: Creates mpy.sampler.BaseSampler

- changed the way VarExpression's (List, Int, Float) store the default
distribution
- simplified sampler to collect default distribution, update the
distribution map for all variables, sample variables from appropriate
distributions, and store the result
- added more test cases in test_sampler to accomodate complex nested
expressions, expression with an input different distribution maps
- enabled Expression class in _expression.py to memoize VarExpressions

Signed-off-by: Akhil Akella <aakella@gpu3.lcrc.anl.gov>
